### ISSUE TITLE
Deleting records from Dataverse database

### ISSUE THREAD

**Issue by philippconzett** [IQSS/dataverse#9832]
We're planning to do some cleanup in the metadata records in the Dataverse database. Some changes we are doing through API, some manually and some through SQL statements. When testing the results of the SQL script containing UPDATE statements, we got a Solr re-indexing failure for a dataset where the value for authorAffiliation in table datasetfieldvalue was updated to NULL. So, it seems set value=NULL results in issues for the value fields in the datasetfieldvalue table for some datasetfieldtypes. 

Instead of UPDATE, I have now created a script using DELETE statements for values we want to remove, and figured out the following:

1. Records in table datasetfieldvalue are children of records in table datasetfield. Therefore, the records in table datasetfieldvalue must be deleted BEFORE the records of table datasetfield are deleted.

2. Records in table datasetfieldcompound must not be deleted, because none of the items that need to be deleted in our cleanup job are the only fields in the compound metadata field. Deleting the datasetfieldcompound records that correspond to the datasetfieldvalues that need to be deleted would result in other records, which we don't want to delete, lacking their corresponding datasetfieldcompound values.

I ran some examples of the type below in our test environment, and there were no error messages:

```
DELETE FROM datasetfieldvalue WHERE id=102230;
DELETE FROM datasetfield WHERE id=151349;
```

Could some of you database experts give some advice on this? Apart from the fact that such changes of course won't be reflected in the version control information of these datasets, is this SQL DELETE method safe to apply?

Thanks!

**Comment by pdurbin** [comment]
@philippconzett should we have API endpoints for the functionality you want so that you don't need to touch the database directly?

**Comment by philippconzett** [comment]
@pdurbin thanks, yes, I think that might be useful. For the cleanup job at stake, it would be good to get some advice whether deleting these records will break anything. Based on my testing, it doesn't break the database, and the datasets still look fine. But I'm wondering whether deleting records might have some consequences for other parts of the application.

**Comment by qqmyers** [comment]
Are you removing things from the latest published version? If so, would the superuser capability to update the current version help? (UI and API). You should be able to delete a value and then republish the existing version. It wouldn't work on older versions though.

**Comment by donsizemore** [comment]
@philippconzett It sounds like what you want to do can't be accomplished with https://guides.dataverse.org/en/latest/api/native-api.html#update-metadata-for-a-dataset ? (remember superUsers can do an update on published datasets without bumping the version). There should be database constraints in place to prevent problems, but as Phil says API endpoints would be much safer.

**Comment by philippconzett** [comment]
@qqmyers @donsizemore  Thanks, I'm using both approaches (GUI and API), in addition to SQL.
 
Ideally, I'd like to use API for all changes to make sure the changes are recorded in version control, but this would be quite time-consuming to do in cases where, e.g., only one instance of a repeatable field needs to be updated, let's say only one of ten author names in a dataset. In these cases, I would have to upload a json file that includes all author names. We have hundreds of cases where personal names are not inverted. With SQL UPDATE commands these can easily be changed.

Of the about 3000 changes, 2500+ are SQL UPDATE statements; about 150 I did manually, because they were complex (e.g. adding keywords in separate fields that currently are placed in one single field); about 200 I'm doing through API, and they are all about adding missing fields/values (I guess there is no safe way to add thing with SQL statements). And in about 150 cases, I need to remove values/fields. The removing changes apply all to the latest version, either draft or latest published version. The SQL DELETE commands seem straight forward. For API, I would need to download the json of all involved dataset, and then remove the fields that shouldn't be there.

@donsizemore  By "database constraints" you mean error messages like the one you get when you try to delete a record that has child records that are related with foreign keys? If so, I don't get any of these because I delete all the children (datasetfieldvalue) before the parents (datasetfield).

@donsizemore "... without bumping the version" --- Are you refering to setting type=updatecurrent when publishing the new version? How does this work? Will there still be a "publication process" involved, or will the new draft just replace the current published version without any publishing involved?

**Comment by donsizemore** [comment]
@philippconzett yes and yes, and... I was going to refer you to Julian by way of support ticket, but he already "eyeballed" your comment above. I do not envy you this task.

**Comment by jggautier** [comment]
Yeah and I mentioned this GitHub issue in [a recent conversation](https://groups.google.com/u/1/g/dataverse-community/c/WY4l0t3qVBs) in the Dataverse Community Google Group. Also going to link this to another GitHub issue at https://github.com/IQSS/dataverse/issues/6362.

**Comment by philippconzett** [comment]
Thanks, @donsizemore and @jggautier. Brief heads-up: I'm working on an API-based script, hopefully to be finished later this week.

**Comment by DS-INRAE** [comment]
Hi @philippconzett did you manage to address the issue with a script? Do we have a place where we could share such scripts that may be useful for other installations? That would be interesting to have in the documentation or GDCC.

**Comment by philippconzett** [comment]
Hi @DS-INRAE, not with a single script. This was a comprehensive work including a combination of scripts with SQL queries, API calls, jq commands, Linux commands and manual tasks in spreadsheets. Altogether, we edited the value in 2225 metadata fields across 733 datasets. I don't think the scripts in the current form are of much value for others without adding extensive explanation. But I'd be happy to discuss (parts of) the clean-up work on shared screen.
